contracts/pcv/PCVDepositAggregator.sol

```diff
@@ -0,0 +1,365 @@
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+contract PCVDepositAggregator is IPCVDepositAggregator, CoreRef {
+    using EnumerableSet for EnumerableSet.AddressSet;
+    using SafeERC20 for IERC20;
+
+    // ---------- Events ----------
+
+    // ---------- Properties ------
+
+    struct PCVDepositInfo {
+        uint128 weight;
+        bool depositsPaused;
+    }
+
+    EnumerableSet.AddressSet private pcvDepositAddresses;
+    
+    mapping(address => PCVDepositInfo) public pcvDepositInfos;
+    
+    uint128 public bufferWeight;
+    uint128 public totalWeight; 
+    address public token;
+    address public override rewardsAssetManager;
+
+    constructor(
+        address _core,
+        address _rewardsAssetManager,
+        address _token,
+        address[] memory _initialPCVDepositAddresses,
+        uint128[] memory _initialPCVDepositWeights,
+        uint128 _bufferWeight
+    ) CoreRef(_core) {
+        if (_initialPCVDepositAddresses.length != _initialPCVDepositWeights.length) {
+            revert("Addresses and weights are not the same length!");
+        }
+
+        if (_rewardsAssetManager == address(0x0)) {
+            revert("Rewards asset manager cannot be null");
+        }
+
+        rewardsAssetManager = _rewardsAssetManager;
+        token = _token;
+        bufferWeight = _bufferWeight;
+        totalWeight = bufferWeight;
+
+        _setContractAdminRole(keccak256("PCV_CONTROLLER_ROLE"));
+
+        for (uint i=0; i < _initialPCVDepositAddresses.length; i++) {
+            _addPCVDeposit(_initialPCVDepositAddresses[i], _initialPCVDepositWeights[i]);
+        }
+    }
+
+    // ---------- Public Functions -------------
+
+    function rebalance() external virtual override {
+        _rebalance();
+    }
+
+    function rebalanceSingle(address pcvDeposit) external virtual override {
+        _rebalanceSingle(pcvDeposit);
+    }
+
+    function deposit() external virtual override {
+        // no-op
+    }
+
+    /**
+     * @notice withdraws the specified amount of tokens from the contract
+     * @dev the implementation is as follows:
+     * 1. check if the contract has enough in the buffer to cover the withdrawal. if so, just use this
+     * 2. if it doesn't, scan through each of the pcv deposits and withdraw from them their overage amounts,
+     *    up to the total amount needed (less the amount already in the buffer)
+     */
+    function withdraw(address to, uint256 amount) external virtual override onlyPCVController {
+        uint totalBalance = getTotalBalance();
+
+        if (amount > totalBalance) {
+            revert("Not enough balance to withdraw");
+        }
+
+        // If our buffer is full enough, just transfer from that
+        if (IERC20(token).balanceOf(address(this)) >= amount) {
+            IERC20(token).safeTransfer(to, amount);
+        } else {
+            // We're going to have to pull from underlying deposits
```

Comment:
What about if we do all the overages and thats not enough? This method should not fail under any circumstance if possible

Reply:
Yep, you found a bug. Fixing.

Reply:
Fixed.